scanner/enricher/nvd/nvd.go

```diff
@@ -0,0 +1,421 @@
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/facebookincubator/nvdtools/cveapi/nvd/schema"
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/libvuln/driver"
+	"github.com/quay/zlog"
+	"github.com/stackrox/rox/pkg/utils"
+)
+
+var (
+	_ driver.Enricher          = (*Enricher)(nil)
+	_ driver.EnrichmentUpdater = (*Enricher)(nil)
+
+	defaultFeed *url.URL
+)
+
+const (
+	// Type is the type of data returned from the Enricher's Enrich method.
+	Type = `message/vnd.stackrox.scannerv4.vulnerability; enricher=nvd schema=https://csrc.nist.gov/schema/nvd/api/2.0/source_api_json_2.0.schema`
+
+	// DefaultFeeds is the default place to look for CVE feeds.
+	DefaultFeeds = `https://services.nvd.nist.gov/rest/json/cves/2.0/`
+
+	// This appears above and must be the same.
+	name = `nvd`
+
+	// First year for the yearly CVE feeds: https://nvd.nist.gov/vuln/data-feeds
+	firstYear = 2002
+)
+
+func init() {
+	var err error
+	defaultFeed, err = url.Parse(DefaultFeeds)
+	utils.CrashOnError(err)
+}
+
+// Enricher provides NVD CVE data as enrichments to a VulnerabilityReport.
+//
+// Configure must be called before any other methods.
+type Enricher struct {
+	driver.NoopUpdater
+	c            *http.Client
+	feed         *url.URL
+	apiKey       string
+	callInterval time.Duration
+}
+
+// Config is the configuration for Enricher.
+type Config struct {
+	FeedRoot     *string `json:"feed_root" yaml:"feed_root"`
+	APIKey       *string `json:"api_key" yaml:"api_key"`
+	CallInterval *string `json:"call_interval" yaml:"call_interval"`
+}
+
+// NewFactory creates a Factory for the NVD enricher.
+func NewFactory() driver.UpdaterSetFactory {
+	set := driver.NewUpdaterSet()
+	_ = set.Add(&Enricher{})
+	return driver.StaticSet(set)
+}
+
+// Configure implements driver.Configurable.
+func (e *Enricher) Configure(ctx context.Context, f driver.ConfigUnmarshaler, c *http.Client) error {
+	ctx = zlog.ContextWithValues(ctx, "component", "enricher/nvd/Enricher/Configure")
+	var cfg Config
+	e.c = c
+	if err := f(&cfg); err != nil {
+		return err
+	}
+	if cfg.APIKey != nil {
+		e.apiKey = *cfg.APIKey
+	}
+	// NVD limits users without an API key to roughly one call every 6 seconds.
+	// With an API key, it is roughly one call every 0.6 seconds.
+	// We'll play it safe and do one call every 3 seconds.
+	// As of writing there are ~216,000 vulnerabilities, so this whole process should take ~5.4 minutes.
+	e.callInterval = 3 * time.Second
+	if cfg.CallInterval != nil {
+		var err error
+		e.callInterval, err = time.ParseDuration(*cfg.CallInterval)
+		if err != nil {
+			return err
+		}
+	}
+	if cfg.FeedRoot != nil {
+		if !strings.HasSuffix(*cfg.FeedRoot, "/") {
+			return fmt.Errorf("URL missing trailing slash: %q", *cfg.FeedRoot)
+		}
+		u, err := url.Parse(*cfg.FeedRoot)
+		if err != nil {
+			return err
+		}
+		e.feed = u
+	} else {
+		var err error
+		e.feed, err = defaultFeed.Parse(".")
```

Comment:
what does this mean?

Reply:
https://pkg.go.dev/net/url#Parse

Same as:

```
		u := *defaultFeed
		e.feed = &u 
```